### PR TITLE
MAINT: Move spline interpolation code to spline.c

### PR DESCRIPTION
**Pull request by jaimefrio**
This PR moves some of the code in `ndimage/ni_interpolation.c` to a new file `ni_splines.c`. The main purpose of the changes are documentary: the new code is profusely documented, and cites a couple of the original articles on which it is based. This is a first step in finding a path forward for #8465, namely making the code more understandable.

Some of the refactorings may have some marginal performance gains, e.g. the calculation of the spline interpolation weights should be faster, since it has less branches.

### CODE REVIEW

**Comment by rgommers** [comment]
I read through the changes, looks like a major improvement in readability. I followed how the big chunks of code were moved, but didn't verify all refactoring in detail. I'm inclined to rely on the tests for that and just hit the green button. Is there anything in particular that needs a closer review?

**Comment by jaimefrio** [comment]
The code is heavily, let's call it *inspired*, on the [reference implementation](http://bigwww.epfl.ch/thevenaz/interpolation/) from Dr. Philippe Thévenaz, a researcher in Prof. Michael Unser's [group](http://bigwww.epfl.ch/people.html), so I'm pretty confident the general approach is sound.

Judging by some of the constructs in the existing ndimage code, I'm inclined to think that either the existing ndimage code was also based in an earlier version of the same reference implementation, or they both share a common ancestor.

Let me self-review the code again, and see if I can add a few more tests, and I'll ping here when I'm 100% happy with this.

**Comment by jaimefrio** [comment]
OK, have added some tests to check that the result of spline filtering is the same you'd get from computing the spline coefficients solving a linear system of equations. If it passes all the CI tests, I'm happy with the state of this, can be merged.

File: scipy/ndimage/tests/test_splines.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        for idx, knot_value in enumerate(knot_values[row + 1:]):
+            matrix[row, start + step*idx] += knot_value
+            matrix[-row - 1, -start - 1 - step*idx] += knot_value
+
+    return matrix / knot_values_sum
+
+
+@pytest.mark.parametrize('order', [0, 1, 2, 3, 4, 5])
+def test_spline_filter_vs_matrix_solution(order):
+    n = 100
```
Comment (charris) [inline scipy/ndimage/tests/test_splines.py]: Can also use `eye` and filter along an axis, the result should then match the matrix inverse.

Comment (charris) [reply scipy/ndimage/tests/test_splines.py]: That is a more complete test.

Comment (charris) [reply scipy/ndimage/tests/test_splines.py]: The first axis, to be precise.

Comment (jaimefrio) [reply scipy/ndimage/tests/test_splines.py]: It is indeed more complete, have written it as matrix product being equal to the identity, PTAL.

**Comment by charris** [comment]
The test failures look unrelated.

File: scipy/ndimage/tests/test_splines.py (hunk 1 above)
Comment (charris) [reply scipy/ndimage/tests/test_splines.py]: Looks good. Might get a bit more completeness by testing the filtering along both axis just to make sure the `axis` keyword works :)

**Review by charris — APPROVED** [review]

**Comment by jaimefrio** [comment]
Have added the extra test over the other axis.

**Comment by jaimefrio** [comment]
The failures seem unrelated to this PR.

**Comment by charris** [comment]
Thanks Jaime.

**Pull request merged**